File: src/main/java/org/jcodec/common/AudioFormat.java
package org.jcodec.common;

/**
 * This class is part of JCodec ( www.jcodec.org ) This software is distributed
 * under FreeBSD License
 * 
 * @author The JCodec project
 * 
 */
public class AudioFormat {

    private int sampleRate;
    private int sampleSizeInBits;
    private int channelCount;
    private boolean signed;
    private boolean bigEndian;

    // Common audio formats
    public static AudioFormat STEREO_48K_S16_BE = new AudioFormat(48000, 16, 2, true, true);
    public static AudioFormat STEREO_48K_S16_LE = new AudioFormat(48000, 16, 2, true, false);

    public static AudioFormat STEREO_48K_S24_BE = new AudioFormat(48000, 24, 2, true, true);
    public static AudioFormat STEREO_48K_S24_LE = new AudioFormat(48000, 24, 2, true, false);

    public static AudioFormat MONO_48K_S16_BE = new AudioFormat(48000, 16, 1, true, true);
    public static AudioFormat MONO_48K_S16_LE = new AudioFormat(48000, 16, 1, true, false);

    public static AudioFormat MONO_48K_S24_BE = new AudioFormat(48000, 24, 1, true, true);
    public static AudioFormat MONO_48K_S24_LE = new AudioFormat(48000, 24, 1, true, false);

    public static AudioFormat STEREO_44K_S16_BE = new AudioFormat(44100, 16, 2, true, true);
    public static AudioFormat STEREO_44K_S16_LE = new AudioFormat(44100, 16, 2, true, false);

    public static AudioFormat STEREO_44K_S24_BE = new AudioFormat(44100, 24, 2, true, true);
    public static AudioFormat STEREO_44K_S24_LE = new AudioFormat(44100, 24, 2, true, false);

    public static AudioFormat MONO_44K_S16_BE = new AudioFormat(44100, 16, 1, true, true);
    public static AudioFormat MONO_44K_S16_LE = new AudioFormat(44100, 16, 1, true, false);

    public static AudioFormat MONO_44K_S24_BE = new AudioFormat(44100, 24, 1, true, true);
    public static AudioFormat MONO_44K_S24_LE = new AudioFormat(44100, 24, 1, true, false);

    public AudioFormat(int sampleRate, int sampleSizeInBits, int channelCount, boolean signed, boolean bigEndian) {
        this.sampleRate = sampleRate;
        this.sampleSizeInBits = sampleSizeInBits;
        this.channelCount = channelCount;
        this.signed = signed;
        this.bigEndian = bigEndian;
    }

    public AudioFormat(AudioFormat format, int newSampleRate) {
        this(format);
        this.sampleRate = newSampleRate;
    }

    public AudioFormat(AudioFormat format) {
        this(format.sampleRate, format.sampleSizeInBits, format.channelCount, format.signed, format.bigEndian);
    }

    public int getChannels() {
        return channelCount;
    }

    public int getSampleSizeInBits() {
        return sampleSizeInBits;
    }

    public int getSampleRate() {
        return sampleRate;
    }

    public short getFrameSize() {
        return (short) ((sampleSizeInBits >> 3) * channelCount);
    }

    public int getFrameRate() {
        return sampleRate;
    }

    public boolean isBigEndian() {
        return bigEndian;
    }

    public boolean isSigned() {
        return signed;
    }

    public int bytesToSamples(int bytes) {
        return bytes / (channelCount * sampleSizeInBits >> 3);
    }

    public int samplesToBytes(int samples) {
        return samples * (channelCount * sampleSizeInBits >> 3);
    }
}

File: src/main/java/org/jcodec/common/AudioFormat.kt
package org.jcodec.common

/**
 * This class is part of JCodec ( www.jcodec.org ) This software is distributed
 * under FreeBSD License
 *
 * @author The JCodec project
 */
data class AudioFormat(val sampleRate: Int, val sampleSizeInBits: Int, val channels: Int, val isSigned: Boolean, val isBigEndian: Boolean) {

    val frameSize: Short
        get() = ((sampleSizeInBits shr 3) * channels).toShort()

    val frameRate: Int
        get() = sampleRate

    fun bytesToFrames(bytes: Int): Int {
        return bytes / (channels * sampleSizeInBits shr 3)
    }

    fun framesToBytes(samples: Int): Int {
        return samples * (channels * sampleSizeInBits shr 3)
    }

    fun bytesToSamples(bytes: Int): Int {
        return bytes / (sampleSizeInBits shr 3)
    }

    fun samplesToBytes(samples: Int): Int {
        return samples * (sampleSizeInBits shr 3)
    }

    companion object {
        // Common audio formats
        var STEREO_48K_S16_BE = AudioFormat(48000, 16, 2, true, true)
        var STEREO_48K_S16_LE = AudioFormat(48000, 16, 2, true, false)
        var STEREO_48K_S24_BE = AudioFormat(48000, 24, 2, true, true)
        var STEREO_48K_S24_LE = AudioFormat(48000, 24, 2, true, false)
        var MONO_48K_S16_BE = AudioFormat(48000, 16, 1, true, true)
        var MONO_48K_S16_LE = AudioFormat(48000, 16, 1, true, false)
        var MONO_48K_S24_BE = AudioFormat(48000, 24, 1, true, true)
        var MONO_48K_S24_LE = AudioFormat(48000, 24, 1, true, false)
        var STEREO_44K_S16_BE = AudioFormat(44100, 16, 2, true, true)
        var STEREO_44K_S16_LE = AudioFormat(44100, 16, 2, true, false)
        var STEREO_44K_S24_BE = AudioFormat(44100, 24, 2, true, true)
        var STEREO_44K_S24_LE = AudioFormat(44100, 24, 2, true, false)
        var MONO_44K_S16_BE = AudioFormat(44100, 16, 1, true, true)
        var MONO_44K_S16_LE = AudioFormat(44100, 16, 1, true, false)
        var MONO_44K_S24_BE = AudioFormat(44100, 24, 1, true, true)
        var MONO_44K_S24_LE = AudioFormat(44100, 24, 1, true, false)
        fun STEREO_S16_BE(rate: Int): AudioFormat {
            return AudioFormat(rate, 16, 2, true, true)
        }

        fun STEREO_S16_LE(rate: Int): AudioFormat {
            return AudioFormat(rate, 16, 2, true, false)
        }

        fun STEREO_S24_BE(rate: Int): AudioFormat {
            return AudioFormat(rate, 24, 2, true, true)
        }

        fun STEREO_S24_LE(rate: Int): AudioFormat {
            return AudioFormat(rate, 24, 2, true, false)
        }

        fun MONO_S16_BE(rate: Int): AudioFormat {
            return AudioFormat(rate, 16, 1, true, true)
        }

        fun MONO_S16_LE(rate: Int): AudioFormat {
            return AudioFormat(rate, 16, 1, true, false)
        }

        fun MONO_S24_BE(rate: Int): AudioFormat {
            return AudioFormat(rate, 24, 1, true, true)
        }

        fun MONO_S24_LE(rate: Int): AudioFormat {
            return AudioFormat(rate, 24, 1, true, false)
        }

        fun NCH_48K_S16_BE(n: Int): AudioFormat {
            return AudioFormat(48000, 16, n, true, true)
        }

        fun NCH_48K_S16_LE(n: Int): AudioFormat {
            return AudioFormat(48000, 16, n, true, false)
        }

        fun NCH_48K_S24_BE(n: Int): AudioFormat {
            return AudioFormat(48000, 24, n, true, true)
        }

        fun NCH_48K_S24_LE(n: Int): AudioFormat {
            return AudioFormat(48000, 24, n, true, false)
        }

        fun NCH_44K_S16_BE(n: Int): AudioFormat {
            return AudioFormat(44100, 16, n, true, true)
        }

        fun NCH_44K_S16_LE(n: Int): AudioFormat {
            return AudioFormat(44100, 16, n, true, false)
        }

        fun NCH_44K_S24_BE(n: Int): AudioFormat {
            return AudioFormat(44100, 24, n, true, true)
        }

        fun NCH_44K_S24_LE(n: Int): AudioFormat {
            return AudioFormat(44100, 24, n, true, false)
        }

        fun createAudioFormat(format: AudioFormat): AudioFormat {
            return AudioFormat(format.frameRate, format.sampleSizeInBits, format.channels, format.isSigned, format.isBigEndian)
        }

        fun createAudioFormat2(format: AudioFormat, newSampleRate: Int): AudioFormat {
            return format.copy(sampleRate = newSampleRate)
        }
    }

}